sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseReadSplittableDoFn.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.hbase;
+
+import java.util.List;
+import org.apache.beam.sdk.io.hadoop.SerializableConfiguration;
+import org.apache.beam.sdk.io.range.ByteKey;
+import org.apache.beam.sdk.io.range.ByteKeyRange;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.DoFn.BoundedPerElement;
+import org.apache.beam.sdk.transforms.splittabledofn.ByteKeyRangeTracker;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.ConnectionFactory;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+
+/** A SplittableDoFn to read from HBase. */
+@BoundedPerElement
+class HBaseReadSplittableDoFn extends DoFn<Void, Result> {
+  private final SerializableConfiguration serializableConfiguration;
+  private final String tableId;
+  private final SerializableScan scan;
+
+  private Connection connection;
+
+  HBaseReadSplittableDoFn(
+      SerializableConfiguration serializableConfiguration, String tableId, SerializableScan scan) {
+    this.serializableConfiguration = serializableConfiguration;
+    this.tableId = tableId;
+    this.scan = scan;
+  }
+
+  @Setup
+  public void setup() throws Exception {
+    if (connection == null) {
+      connection = ConnectionFactory.createConnection(serializableConfiguration.get());
+    }
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c, ByteKeyRangeTracker tracker) throws Exception {
+    TableName tableName = TableName.valueOf(tableId);
+    Table table = connection.getTable(tableName);
+    final ByteKeyRange range = tracker.currentRestriction();
+    Scan newScan =
```

Comment:
WDYT about making this a `DoFn<Scan, Result>` instead? So that it can be generalized to HBaseIO.readAll() in the future.

Or maybe it should be something more than `Scan` - maybe a tuple of tableId and Scan?

Either way, I'd say that a splittable `DoFn<Void, _>` should be considered an antipattern because it under-uses the composability aspect of SDF.

Reply:
Sure, I had initially done it with `DoFn<Scan, Result>` when I read your blog post but I saw I could hack it with Void, I will change. Now I am curious about your tuple idea. A pity we don't have tuples, maybe KV<Scan, String> ?

Reply:
Actually I ended up creating a new entity `HBaseQuery` because Scan does not contain the tableId, but the DoFn is now consistent with what  you proposed.